dipy/denoise/noise_estimate.py

```diff
@@ -217,53 +223,46 @@ def _piesno_3D(data, N, alpha=0.01, l=100, itermax=100, eps=1e-5,
     K = data.shape[-1]
     sum_m2 = np.sum(data**2, axis=-1, dtype=np.float32)
 
-    sigma = np.zeros(phi.shape, dtype=phi.dtype)
-    mask = np.zeros(phi.shape + data.shape[:-1])
+    sigma_prev = 0
+    sigma = m
+    prev_idx = 0
+    mask = np.zeros(data.shape[:-1], dtype=np.bool)
 
     lambda_minus = _inv_nchi_cdf(N, K, alpha/2)
     lambda_plus = _inv_nchi_cdf(N, K, 1 - alpha/2)
 
-    pos = 0
-    max_length_omega = 0
 
-    for num, sig in enumerate(phi):
+    for sigma_init in phi:
 
-        sig_prev = 0
-        omega_size = 1
-        idx = np.zeros(sum_m2.shape, dtype=np.bool)
+        s = sum_m2 / (2 * K * sigma_init**2)
+        found_idx = np.sum(np.logical_and(lambda_minus <= s, s <= lambda_plus), dtype=np.int16)
 
-        for n in range(itermax):
+        if found_idx > prev_idx:
+            sigma = sigma_init
+            prev_idx = found_idx
 
-            if np.abs(sig - sig_prev) < eps:
-                break
+    for n in range(itermax):
 
-            s = sum_m2 / (2 * K * sig**2)
-            idx = np.logical_and(lambda_minus <= s, s <= lambda_plus)
-            omega = data[idx, :]
+        if np.abs(sigma - sigma_prev) < eps:
+            break
 
-            # If no point meets the criterion, exit
-            if omega.size == 0:
-                omega_size = 0
-                break
+        s = sum_m2 / (2 * K * sigma**2)
+        mask[...] = np.logical_and(lambda_minus <= s, s <= lambda_plus)
+        omega = data[mask, :]
 
-            sig_prev = sig
+        # If no point meets the criterion, exit
+        if omega.size == 0:
+            break
```

Comment:
Yes thats the latest commit I added
On Dec 13, 2015 02:44, "Eleftherios Garyfallidis" <notifications@github.com>
wrote:

> In dipy/denoise/noise_estimate.py
> <https://github.com/nipy/dipy/pull/745#discussion_r47439250>:
>
> >
> > -            sig_prev = sig
> > +        # If no point meets the criterion, exit
> > +        if omega.size == 0:
> > +            break
>
> @samuelstjean <https://github.com/samuelstjean> and @arokem
> <https://github.com/arokem> is this now tested?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/745/files#r47439250>.
>